lib/fluent/plugin/out_geoip.rb

```diff
@@ -102,82 +56,7 @@ def shutdown
 
   def write(chunk)
     chunk.msgpack_each do |tag, time, record|
-      Fluent::Engine.emit(tag, time, add_geoip_field(record))
-    end
-  end
-
-  private
-
-  def json?(text)
-    text.match(/^\[.+\]$/) || text.match(/^\{.+\}$/)
-  end
-
-  def quoted_value?(text)
-    # to improbe compatibility with fluentd v1-config
-    trim_quote = text[1..text.size-2]
-    text.match(/(^'.+'$|^".+"$)/)
-  end
-
-  def add_geoip_field(record)
-    placeholder = create_placeholder(geolocate(get_address(record)))
-    return record if @skip_adding_null_record && placeholder.values.first.nil?
-    @map.each do |record_key, value|
-      if value.match(REGEXP_PLACEHOLDER_SINGLE)
-        rewrited = placeholder[value]
-      elsif json?(value)
-        rewrited = value.gsub(REGEXP_PLACEHOLDER_SCAN) {|match|
-          match = match[1..match.size-2] if quoted_value?(match)
-          Yajl::Encoder.encode(placeholder[match])
-        }
-        rewrited = parse_json(rewrited)
-      else
-        rewrited = value.gsub(REGEXP_PLACEHOLDER_SCAN, placeholder)
-      end
-      record.store(record_key, rewrited)
-    end
-    return record
-  end
-
-  def parse_json(message)
-    begin
-      return Yajl::Parser.parse(message)
-    rescue Yajl::ParseError => e
-      log.info "geoip: failed to parse '#{message}' as json.", :error_class => e.class, :error => e.message
-      return nil
-    end
-  end
-
-  def get_address(record)
-    address = {}
-    @geoip_lookup_key.each do |field|
-      address.store(field, record[field]); next if not record[field].nil?
-      key = field.split('.')
-      obj = record
-      key.each {|k|
-        break obj = nil if not obj.has_key?(k)
-        obj = obj[k]
-      }
-      address.store(field, obj)
-    end
-    return address
-  end
-
-  def geolocate(addresses)
-    geodata = {}
-    addresses.each do |field, ip|
-      geo = ip.nil? ? nil : @geoip.look_up(ip)
-      geodata.store(field, geo)
-    end
-    return geodata
-  end
-
-  def create_placeholder(geodata)
-    placeholder = {}
-    @placeholder_keys.each do |placeholder_key|
-      position = placeholder_key.match(REGEXP_PLACEHOLDER_SINGLE)
-      next if position.nil? or geodata[position[:record_key]].nil?
-      placeholder.store(placeholder_key, geodata[position[:record_key]][position[:geoip_key].to_sym])
+      Fluent::Engine.emit(tag, time, @supplement.add_geoip_field(record))
```

Comment:
`Fluent::Engine.emit` can replace `router.emit` in fluentd 0.10.58 or later.
see: https://github.com/fluent/fluentd/blob/v0.10/ChangeLog#L40